src/main.cpp

```diff
@@ -3446,131 +3500,104 @@ bool ActivateBestChain(CValidationState &state, const CChainParams &chainparams,
+        // extend this chain instead.
+        if (!fOneDone && PV && PV->IsReorgInProgress())
         {
-            // BU move to activatebestchainstep  uiInterface.NotifyBlockTip(fInitialDownload, pindexNewTip);
-
-            if (!IsInitialBlockDownload())
+            // find out if this block and chain are more work than the chain
+            // being reorg'd to.  If not then just return.  If so then kill the reorg and
+            // start connecting this chain.
+            if (pindexMostWork->nChainWork > PV->MaxWorkChainBeingProcessed())
             {
-                // Find the hashes of all blocks that weren't previously in the best chain.
-                std::vector<uint256> vHashes;
-                CBlockIndex *pindexToAnnounce = pindexNewTip;
-                while (pindexToAnnounce != pindexFork)
-                {
-                    vHashes.push_back(pindexToAnnounce->GetBlockHash());
-                    pindexToAnnounce = pindexToAnnounce->pprev;
-                    if (vHashes.size() == MAX_BLOCKS_TO_ANNOUNCE)
-                    {
-                        // Limit announcements in case of a huge reorganization.
-                        // Rely on the peer's synchronization mechanism in that case.
-                        break;
-                    }
-                }
-                // Relay inventory, but don't relay old inventory during initial block download.
-                int nBlockEstimate = 0;
-                if (fCheckpointsEnabled)
-                    nBlockEstimate = Checkpoints::GetTotalBlocksEstimate(chainparams.Checkpoints());
-                {
-                    LOCK(cs_vNodes);
-                    BOOST_FOREACH (CNode *pnode, vNodes)
-                    {
-                        if (chainActive.Height() >
-                            (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : nBlockEstimate))
-                        {
-                            BOOST_REVERSE_FOREACH (const uint256 &hash, vHashes)
-                            {
-                                pnode->PushBlockHash(hash);
-                            }
-                        }
-                    }
-                }
-                // BU: commented out and added in ActivateBestChainStep to support PV
-                // Notify external listeners about the new tip.
-                // if (!vHashes.empty()) {
-                // GetMainSignals().UpdatedBlockTip(pindexNewTip);
-                //}
+                // kill all validating threads except our own.
+                boost::thread::id this_id(boost::this_thread::get_id());
+                PV->StopAllValidationThreads(this_id);
             }
+            else
+                return true;
         }
-    } while (pindexMostWork > chainActive.Tip());
+
+        if (!ActivateBestChainStep(state, chainparams, pindexMostWork,
+                pblock && pblock->GetHash() == pindexMostWork->GetBlockHash() ? pblock : NULL, fParallel))
```

Comment:
while some argue that basic order of operations should be known, I think that it would improve code clarity to add parens in this case since the relative order of ? vs && is pretty esoteric.  In other words, is this statement:
(pblock && pblock->GetHash() == pindexMostWork->GetBlockHash()) ? pblock : NULL
or
pblock && (pblock->GetHash() == pindexMostWork->GetBlockHash()) ? pblock : NULL)